src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,491 @@
 
-/// Capacities of each instance of a resource.
-struct ResourceInstanceCapacities {
-  std::vector<FixedPoint> total;
-  std::vector<FixedPoint> available;
-};
+  /// Get the value of a particular resource.
+  /// NOTE: the resource MUST already exist in this set, otherwise a check
+  /// fail will occur.
+  FixedPoint Get(ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    RAY_CHECK(ptr) << "Resource not found: " << resource_id;
+    return *ptr;
+  }
 
-// Data structure specifying the capacity of each resource requested by a task.
-class ResourceRequest {
- public:
-  /// List of predefined resources required by the task.
-  std::vector<FixedPoint> predefined_resources;
-  /// List of custom resources required by the task.
-  absl::flat_hash_map<int64_t, FixedPoint> custom_resources;
+  /// Get the value of a particular resource. If the resource doesn't exist, return 0.
+  FixedPoint GetOrZero(ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    if (ptr == nullptr) {
+      return FixedPoint(0);
+    } else {
+      return *ptr;
+    }
+  }
+
+  /// Set a resource to the given value.
+  /// NOTE: if the new value is 0, the resource will be removed.
+  ResourceRequest &Set(ResourceID resource_id, FixedPoint value) {
+    if (value == 0) {
+      if (IsPredefinedResource(resource_id)) {
+        predefined_resources_[resource_id.ToInt()] = 0;
+      } else {
+        custom_resources_.erase(resource_id.ToInt());
+      }
+    } else {
+      auto ptr = GetPointer(resource_id);
+      if (ptr == nullptr) {
+        custom_resources_[resource_id.ToInt()] = value;
+      } else {
+        *ptr = value;
+      }
+    }
+    return *this;
+  }
+
+  /// Check whether a particular resource exist.
+  bool Has(ResourceID resource_id) const { return GetOrZero(resource_id) > 0; }
+
+  /// Clear the whole set.
+  void Clear() {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      predefined_resources_[i] = 0;
+    }
+    custom_resources_.clear();
+  }
+
+  /// Cap the resource values in this set by those in another set.
+  void Cap(const ResourceRequest &other) {
+    for (auto resource_id : ResourceIds()) {
+      auto this_value = Get(resource_id);
+      auto other_value = other.GetOrZero(resource_id);
+      if (this_value > other_value) {
+        Set(resource_id, other_value);
+      }
+    }
+  }
+
+  /// Remove the negative values in this set.
+  void RemoveNegative() {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      if (predefined_resources_[i] < 0) {
+        predefined_resources_[i] = 0;
+      }
+    }
+    for (auto it = custom_resources_.begin(); it != custom_resources_.end();) {
+      if (it->second < 0) {
+        custom_resources_.erase(it++);
+      } else {
+        it++;
+      }
+    }
+  }
+
+  /// Return the number of resources in this set.
+  size_t Size() const { return ResourceIds().size(); }
```

Comment:
let's make this O(1)?

Reply:
I can't make it fully O(1). Because we need to filter out the 0-value resources. 
I've make the "custom_resources" to remove 0-value resources automatically. But it's not possible for predefined_resources. 
After the latest optimization, the complexity is now O(len(predefined_resources) + 1).